gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,432 @@
+    "AU",
+    "TR",
+    "BE",
+    "NI",
+    "DK",
+    "AT",
+    "SG",
+    "IS",
+    "KR",
+    "BA",
+    "UY",
+    "IL",
+    "CA",
+    "RS",
+    "CZ",
+]
+
+
+class EntsoePandasClientAdapter(EntsoePandasClient):
+    def __init__(self, api_key: str = None):
+        super().__init__(api_key=api_key)
+
+    def _base_request(
+        self, params: Dict, start: pd.Timestamp, end: pd.Timestamp
+    ) -> requests.Response:
+        """Overwrites EntsoePandasClient _base_request method"""
+        start_str = self._datetime_to_str(start)
+        end_str = self._datetime_to_str(end)
+        base_params = {
+            "securityToken": self.api_key,
+            "periodStart": start_str,
+            "periodEnd": end_str,
+        }
+        params.update(base_params)
+        response = self.session.get(url=ENTSOE_URL, params=params, proxies=None, timeout=None)
+        return response
+
+    def _query_generation_per_plant(
+        self, country_code: Union[Area, str], start: pd.Timestamp, end: pd.Timestamp
+    ):
+        """Overwrites EntsoePandasClient query_generation_per_plant method"""
+        area = lookup_area(country_code)
+        params = {
+            "documentType": "A73",
+            "processType": "A16",
+            "in_Domain": area.code,
+        }
+        response = self._base_request(params=params, start=start, end=end)
+        text = response.text
+        if "Unauthorized" in text:
+            raise ValueError("Invalid API key")
+
+        df_generation = parse_generation(text, per_plant=True, include_eic=False)
+
+        # check if empty
+        if df_generation.empty:
+            raise ValueError("The Entsoe dataframe is empty. Please check the query parameters.")
+        # localize or covert to UTC+0
+        if df_generation.index.tzinfo is not None:
+            df_generation.index = df_generation.index.tz_convert("UTC")
+        else:
+            df_generation.index = df_generation.index.tz_localize("UTC")
+
+        # df_generation is a MultiIndex DataFrame
+        df_generation.columns = df_generation.columns.set_levels(
+            df_generation.columns.levels[0].str.encode("latin-1").str.decode("utf-8"), level=0
+        )
+        # Truncation will fail if data is not sorted along the index
+        df_generation = df_generation.sort_index(axis=0)
+        if df_generation.columns.nlevels == 2:
+            df_generation = (
+                df_generation.assign(newlevel="Actual Aggregated")
+                .set_index("newlevel", append=True)
+                .unstack("newlevel")
+            )
+        df_generation = df_generation.loc[start:end]
+
+        return df_generation
+
+
+class CarbonEmissionsHandler:
+    """The most recent Enstsoe-y version (v0.6.16) is only compatible with python 3.9.
+    Therefore, I am using v0.4.4 whose URL does not work. So, I am overwriting some methods
+    from https://github.com/EnergieID/entsoe-py (branch v0.4.4) to make this work.
+    """
```

Comment:
This comment should be in EntsoePandasClientAdapter right? 

> The most recent Enstsoe-y version (v0.6.16) is only compatible with python 3.9

Might it be, that the raw client does not suffer from this limitation? Maybe only pandas expecting 3.11?